js/subprocess.ts

```diff
@@ -0,0 +1,136 @@
+import * as dispatch from "./dispatch";
+import * as flatbuffers from "./flatbuffers";
+import * as msg from "gen/msg_generated";
+import { assert, unreachable } from "./util";
+import { close, File } from "./files";
+import { ReadCloser, WriteCloser } from "./io";
+
+/** How to handle subsubprocess stdio.
+ *
+ * "inherit" The default if unspecified. The child inherits from the
+ * corresponding parent descriptor.
+ *
+ * "piped"  A new pipe should be arranged to connect the parent and child
+ * subprocesses.
+ *
+ * "null" This stream will be ignored. This is the equivalent of attaching the
+ * stream to /dev/null.
+ */
+export type SubprocessStdio = "inherit" | "piped" | "null";
+
+// TODO Maybe extend VSCode's 'CommandOptions'?
+// tslint:disable-next-line:max-line-length
+// See https://code.visualstudio.com/docs/editor/tasks-appendix#_schema-for-tasksjson
+export interface RunOptions {
+  args: string[];
+  cwd?: string;
+  stdout?: SubprocessStdio;
+  stderr?: SubprocessStdio;
+  stdin?: SubprocessStdio;
+}
+
+export class Subprocess {
+  readonly rid: number;
+  readonly pid: number;
+  readonly stdin?: WriteCloser;
+  readonly stdout?: ReadCloser;
+  readonly stderr?: ReadCloser;
+
+  // @internal
+  constructor(res: msg.RunRes) {
+    this.rid = res.rid();
+    this.pid = res.pid();
+
+    if (res.stdinRid() > 0) {
+      this.stdin = new File(res.stdinRid());
+    }
+
+    if (res.stdoutRid() > 0) {
+      this.stdout = new File(res.stdoutRid());
+    }
+
+    if (res.stderrRid() > 0) {
+      this.stderr = new File(res.stderrRid());
+    }
+  }
+
+  async status(): Promise<SubprocessStatus> {
+    return await runStatus(this.rid);
+  }
+
+  close(): void {
+    close(this.rid);
+  }
+}
+
+export interface SubprocessStatus {
+  success: boolean;
+  code?: number;
+  signal?: number; // TODO: Make this a string, e.g. 'SIGTERM'.
+}
+
+function stdioMap(s: SubprocessStdio): msg.SubprocessStdio {
+  switch (s) {
+    case "inherit":
+      return msg.SubprocessStdio.Inherit;
+    case "piped":
+      return msg.SubprocessStdio.Piped;
+    case "null":
+      return msg.SubprocessStdio.Null;
+    default:
+      return unreachable();
+  }
+}
+
+export function run(opt: RunOptions): Subprocess {
+  const builder = flatbuffers.createBuilder();
+  const argsOffset = msg.Run.createArgsVector(
+    builder,
+    opt.args.map(a => builder.createString(a))
+  );
+  const cwdOffset = opt.cwd == null ? -1 : builder.createString(opt.cwd);
+  msg.Run.startRun(builder);
+  msg.Run.addArgs(builder, argsOffset);
+  if (opt.cwd != null) {
+    msg.Run.addCwd(builder, cwdOffset);
+  }
+  if (opt.stdin) {
+    msg.Run.addStdin(builder, stdioMap(opt.stdin!));
+  }
+  if (opt.stdout) {
+    msg.Run.addStdout(builder, stdioMap(opt.stdout!));
+  }
+  if (opt.stderr) {
+    msg.Run.addStderr(builder, stdioMap(opt.stderr!));
+  }
+  const inner = msg.Run.endRun(builder);
+  const baseRes = dispatch.sendSync(builder, msg.Any.Run, inner);
```

Comment:
Okay *if* we hereby agree that RIDs should be assigned in on the JS side and not on the rust side and we'll do that in the future..
Otherwise this adds a low level API that is fundamentally sync, which I will not approve.

Reply:
Let's discuss fundamentally sync [here](https://github.com/denoland/deno/pull/1156/files#r233328785).

I'm not necessarily against allocating RIDs on the JS side... but I don't understand the benefit. The table is on the privileged side, so it seems allocating it there makes more sense.